_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +87,141 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
     return true;
 }
 
+class ReferenceFrameStorage
+{
+private:
+    class FrameLocker
+    {
+    private:
+        UMC::FrameMemID m_frameId = -1;
+        std::shared_ptr<mfx_UMC_FrameAllocator> m_frameAllocator;
+        bool ValidFrameId()
+        {
+            return (m_frameId >= 0);
+        }
+
+        void Lock()
+        {
+            if (ValidFrameId())
+            {
+                if (m_frameAllocator->IncreaseReference(m_frameId) != UMC::UMC_OK)
+                {
+                    throw vp9_exception(UMC::UMC_ERR_FAILED);
+                }
+            }
+        }
+
+        void Unlock()
+        {
+            if (ValidFrameId())
+            {
+                if (m_frameAllocator->DecreaseReference(m_frameId) != UMC::UMC_OK)
+                {
+                    throw vp9_exception(UMC::UMC_ERR_FAILED);
+                }
+            }
+        }
+
+        FrameLocker(const FrameLocker &) = delete;
+        FrameLocker(const FrameLocker &&) = delete;
+        FrameLocker& operator=(const FrameLocker &) = delete;
+    public:
+        FrameLocker() = default;
+
+        FrameLocker(UMC::FrameMemID frameId, std::shared_ptr<mfx_UMC_FrameAllocator> &frameAllocator) :
+            m_frameId(frameId), m_frameAllocator(frameAllocator)
+        {
+            if (m_frameAllocator == nullptr)
+            {
+                throw vp9_exception(UMC::UMC_ERR_FAILED);
+            }
+            Lock();
+        }
+
+        FrameLocker& operator=(FrameLocker &&locker)
```

Comment:
U don't actually **move** _m_frameId/m_frameAllocator_ here, _locker_ still holds these values and temporary object will try to _Unlock_ them in its destructor.

Reply:
Fixed.